### PR TITLE
linker: add _image_ram_size symbol to linker symbol definitions

### PR DESCRIPTION
**Pull request by danieldegrasse**
Add _image_ram_size symbol to linker definitions. This symbol
corresponds to the total RAM used by the image.

This PR also adds a usage of `_image_ram_size`. The reason this symbol is introduced is because GCC will not compile the expression `((uint32_t)&_image_ram_end) - ((uint32_t)&_image_ram_start)`, with the error `error: initializer element is not constant`.

### CODE REVIEW

**Review by andyross — DISMISSED** [review]
This PR in particular is harmless, being just an optimization on top of an existing API.  But I don't really like the API, FWIW.  Zephyr apps that can assume they run in a contiguous region are becoming a smaller and smaller subset of the platform list.  Seems like the framework needs to grow to keep up.

**Comment by danieldegrasse** [comment]
> This PR in particular is harmless, being just an optimization on top of an existing API. But I don't really like the API, FWIW. Zephyr apps that can assume they run in a contiguous region are becoming a smaller and smaller subset of the platform list. Seems like the framework needs to grow to keep up.

This is a good point- the specific use case this PR enables is executing from a RAM region on the iMX RT series. I think `zephyr_code_relocate` is probably part of what we want to enable an application to have something like a "base" region where the system boots from, and then relocate large portions of the code to different memory regions

File: include/zephyr/linker/ram-end.ld [hunk 1]
```diff
@@ -9,6 +9,7 @@
 	LAST_RAM_ALIGN
 #endif
 	_image_ram_end = .;
+	_image_ram_used = _image_ram_start - _image_ram_end;
```
Comment (nordicjm) [inline include/zephyr/linker/ram-end.ld:12]: this does not give you the used size unless your end address is lower than your start address

**Review by nordicjm — CHANGES_REQUESTED** [review]
should have been a NACK

File: include/zephyr/linker/ram-end.ld (hunk 1 above)
Comment (tejlmand) [reply include/zephyr/linker/ram-end.ld:12]: and please use `size`, to be consistent with other symbol definitions.

File: include/zephyr/linker/linker-defs.h [hunk 2]
```diff
@@ -133,6 +133,7 @@ extern char _flash_used[];
 /* datas, bss, noinit */
 extern char _image_ram_start[];
 extern char _image_ram_end[];
+extern char _image_ram_used[];
```
Comment (tejlmand) [inline include/zephyr/linker/linker-defs.h:136]:
```suggestion
extern char _image_ram_size[];
```
to be aligned with other definitions.

**Review by tejlmand — CHANGES_REQUESTED** [review]

File: include/zephyr/linker/linker-defs.h (hunk 2 above)
Comment (tejlmand) [reply include/zephyr/linker/linker-defs.h:136]: which, based on the commit message was also the original intention, as the commit message says:
> Add _image_ram_size symbol to linker definitions. This symbol
> corresponds to the total RAM used by the image.

File: include/zephyr/linker/ram-end.ld (hunk 1 above)
Comment (danieldegrasse) [reply include/zephyr/linker/ram-end.ld]: > this does not give you the used size unless your end address is lower than your start address

Thank you for catching this- the cast to `uint32_t` was masking this for me when verifying the size calculation was right.

File: include/zephyr/linker/linker-defs.h (hunk 2 above)
Comment (danieldegrasse) [reply include/zephyr/linker/linker-defs.h]: Symbol name is updated to `_image_ram_size`

**Review by mmahadevan108 — DISMISSED** [review]

**Review by tejlmand — DISMISSED** [review]
lgtm, just a minor nit which would be nice to adjust.

File: include/zephyr/linker/ram-end.ld [hunk 3]
```diff
@@ -9,6 +9,7 @@
 	LAST_RAM_ALIGN
 #endif
 	_image_ram_end = .;
+	_image_ram_size = _image_ram_end- _image_ram_start;
```
[the remark this answers is not in the document]
Comment (danieldegrasse) [reply include/zephyr/linker/ram-end.ld]: Updated

**Review by nordicjm — APPROVED** [review]

**Review by mmahadevan108 — APPROVED** [review]

**Review by dcpleung — APPROVED** [review]

**Pull request merged**